src/main/java/frc/robot/subsystems/Lift.java

```diff
@@ -246,9 +255,76 @@ public void initialize() {
             * Constants.Lift.ELEVATOR_MOTOR_ENCODER_DIFFERENCES_SCALAR_INCHES_PER_DEGREE);
   }
 
+  /** Sets the desired position, which the lift may not go to directly. */
+  public void setDesiredPosition(LiftPosition pos) {
+    desiredPosition = pos;
+  }
+
+  /** True if the lift is at the queried position. */
+  private boolean atPosition(LiftPosition positionToCheck) {
+    double armThresholdDegrees =
+        positionToCheck == LiftPosition.STARTING
+            ? Cal.Lift.ARM_START_THRESHOLD_DEGREES
+            : Cal.Lift.ARM_THRESHOLD_DEGREES;
+    double elevatorThresholdInches =
+        positionToCheck == LiftPosition.STARTING
+            ? Cal.Lift.ELEVATOR_START_THRESHOLD_INCHES
+            : Cal.Lift.ELEVATOR_THRESHOLD_INCHES;
+    double elevatorPositionToCheckInches = liftPositionMap.get(positionToCheck).getFirst();
+    double armPositionToCheckDegrees = liftPositionMap.get(positionToCheck).getSecond();
+    double elevatorPositionInches = elevatorEncoder.getPosition();
+    double armPositionDegrees = armEncoder.getPosition();
+    if (Math.abs(armPositionDegrees - armPositionToCheckDegrees) > armThresholdDegrees) {
+      return false;
+    }
+
+    if (Math.abs(elevatorPositionInches - elevatorPositionToCheckInches)
+        > elevatorThresholdInches) {
+      return false;
+    }
+
+    return true;
+  }
+
+  /** Converts a lift position to a relative lift position (above or below starting position). */
+  private static LiftPositionStartRelative getRelativeLiftPosition(LiftPosition pos) {
+    switch (pos) {
+      case STARTING:
+        return LiftPositionStartRelative.AT_START;
+      case SCORE_MID:
+      case SCORE_HIGH:
+      case SHELF:
+        return LiftPositionStartRelative.ABOVE_START;
+      case GRAB_FROM_INTAKE:
+        return LiftPositionStartRelative.BELOW_START;
+      default:
+        // should never be used?
+        return LiftPositionStartRelative.ABOVE_START;
+    }
+  }
+
+  /** Called every scheduler run */
   @Override
   public void periodic() {
-    // This method will be called once per scheduler run
+    // Check if position has updated.
+    if (atPosition(LiftPosition.STARTING)) {
+      latestPosition = LiftPosition.STARTING;
+    } else if (atPosition(desiredPosition)) {
+      latestPosition = desiredPosition;
+    }
+
+    // If the lift is going from below to above or above to below, we have to transit through
+    // starting position. Otherwise, we can go directly to the desired position.
+    LiftPositionStartRelative latestPositionStartRelative = getRelativeLiftPosition(latestPosition);
+    if (getRelativeLiftPosition(desiredPosition) == LiftPositionStartRelative.BELOW_START
```

Comment:
Why do we run `getRelativeLiftPosition(desiredPosition)` in the if statement, but `getRelativeLiftPosition(latestPosition)` is its own variable? I would change the relative desired position into its own variable as well; it's more readable, and it is used multiple times.

Reply:
Changed